src/cryptonote_core/blockchain.cpp

```diff
@@ -3405,6 +3405,113 @@ void Blockchain::flush_invalid_blocks() {
     m_invalid_blocks.clear();
 }
 //------------------------------------------------------------------
+std::vector<eth::bls_public_key> Blockchain::get_removable_nodes() const
+{
+    // TODO: Just calculate an acceleration structure on block receive.
+    std::vector<eth::bls_public_key> bls_pubkeys_in_snl;
+    std::vector<eth::bls_public_key> bls_pubkeys_in_smart_contract;
+    {
+        // NOTE: Extract all the service nodes from the Oxen work-chain _excluding_ those that have
+        // been recently removed (e.g. voluntarily exited or dereg by protocol). Recently removed
+        // nodes on the Oxen side need to be removed from the smart contract side and this and can
+        // be done by aggregating a signature (or without if the wait time has elapsed) from the
+        // network to be removed from the smart contract (which gets re-witnessed by Oxen thereby
+        // removing it from the 'recently_removed_nodes' list).
+        auto sns = service_node_list.get_service_node_list_state();
+        bls_pubkeys_in_snl.reserve(sns.size());
+        for (const auto& sni : sns)
+            bls_pubkeys_in_snl.push_back(sni.info->bls_public_key);
+
+        // NOTE: Because of vote confirmations, we have nodes that can exist in the smart contract
+        // but not the service node list for awhile until they get confirmed at which point they get
+        // added to the SNL.
+        //
+        // We assume that these nodes will be added to the SNL. If it does, it will eventually be
+        // removed from this list and show up in the 'get_service_node_list_state()' branch above.
+        //
+        // If the node gets denied, it will be removed from this list and never added to the SNL.
+        service_node_list.for_each_pending_l2_state(
+            [&bls_pubkeys_in_snl]<typename Event>(const Event& e, const service_nodes::service_node_list::unconfirmed_l2_tx&) {
+                if constexpr  (std::is_same_v<Event, eth::event::NewServiceNode>) {
+                    bls_pubkeys_in_snl.push_back(e.bls_pubkey);
+                } else if constexpr (std::is_same_v<Event, eth::event::ServiceNodeExitRequest>) {
+                } else if constexpr (std::is_same_v<Event, eth::event::ServiceNodeExit>) {
+                } else {
+                    log::error(logcat, "Got unknown event type when iterating each pending L2 state from the SNL!");
+                }
+        });
```

Comment:
All good points. They wouldn't qualify for `liquidation` because they're not in the `recently_removed_nodes` list but they are qualified for `exit` in which case they still have an incentive to do so because the less nodes there are the higher share of the rewards pool they end up with. So all in all still quite problematic.

1 and 2 make me nervous because the delays are effective but only for that duration specified. So things break in additional ways once the network stalls for:

1. 10 minutes or more then new registrations made 10 mins ago will show up in the diff
2. Same as 1 but replace 10 minutes with 24 hours.

Now since the nodes are in the smart contract but not the SNL, they will show up in the `set_difference`. They are now eligible for exiting _but_ not liquidation (because they aren't in the `recently_removed_nodes` and hence don't have a `liquidation_height` defined for nodes to permit liquidating it).

To me that makes the guarantees 1 and 2 provide too weak for me. 3 on the other hand isn't affected by these problems. In a network stalling situation, in 3, the newly registered nodes will never be able to be booted off because Oxen is not witnessing new registrations. This doesn't seem like a problem, since they never enter the SNL they're not in a swarm and don't actually do anything. They won't accrue rewards either since blocks are paused. 

In the case of 3, if the network is _not_ stalled then I don't think the downside is too severe. The bad nodes will be online for 3-4 minutes before they get booted off once the network has caught up to the node's contract ID (e.g. when they observe the registration of the node)? 

--

At a larger scale the core issue I see is that we're trying to reconcile up-to-date contract information from `allServiceNodeIDs` against an out of date Oxen blockchain. I think the suggestion below to lock-step the L2 update makes sure that all L2 nodes diff the SNL and smart contract with the same data from the correct time-frame is key.

4. What if we just snapped a list of `allServiceNodeIDs` at the L2 `safe_height` when the L2 tracker is updating and returning that list when it's requested? If the network is stalled and has gone completely out of the safe-window, if we didn't succesfully snap the list at that time, I'd say we just fallback to refusing all exits and liquidations until the L2 provider (as per 3.) catches up as I believe the `safe_height` will just teleport to the latest once its gets there.

If the network is slightly behind well since everyone's snapped the list then you either have it or you don't, (maybe your L2 provider lagged behind and jumped over it). In which case you become a non-signer, but that'd universally be the case for all approaches.

Side bonus that since we use the L2 height that's committed into the block, if you wanted to, you could vet the SNL activity against the L2 and you could do so reliably given this information. Having the contract list and SNL be synced by the L2 safe height lets the rest of the codebase reasonably assume that they're always dealing with relevant data than having to check if they need to filter the data to a sane subset or not.